ebean-test/src/test/java/org/tests/batchinsert/TestBatchInsertFlush.java

```diff
@@ -359,4 +369,79 @@ public void testFlushOnSetProperty() {
       txn.end();
     }
   }
+
+  @Test
+  public void testFlushWithOrder() {
+
+    ResetBasicData.reset();
+
+    try (Transaction txn = DB.beginTransaction(TxScope.requiresNew())) {
+      // do anything, that enables batch mode
+      // in ebean 14
+      DB.insertAll(List.of(new Acl("test")));
+      // does the same as txn.setBatchMode(true);
+
+      // Save contact without group assigned
+      Contact contact = new Contact();
+      contact.setCustomer(DB.find(Customer.class).setMaxRows(1).findOne());
+      DB.save(contact);
+
+      // create new group with fixed id
+      ContactGroup group = new ContactGroup();
+      group.setName("Max");
+      group.setId(42);
+      DB.save(group);
+
+      // assign new group to contact
+      contact.setGroup(group);
+      DB.save(contact);
+      // Ebean holds in it's internal batch queue Contact:0. ContactGroup:1
+
+      txn.commit();
+    }
+  }
+
+
+  @Test
+  public void testWithBatch() {
+
+    ResetBasicData.reset();
+
+    Map<Integer, String> contactCsv = Map.of(123, "Noemi", 124, "Roland");
+    Map<Integer, String> contactGroupCsv = Map.of(42, "Praml", 43, "Max");
+    Map<Integer, Integer> mappingCsv = Map.of(123, 42);
+    try (Transaction txn = DB.beginTransaction(TxScope.requiresNew())) {
+      txn.setBatchMode(true);
+
+      Customer someCustomer = DB.find(Customer.class).setMaxRows(1).findOne(); // just to get no error
+      // Save contact without group assigned
+
+      Map<Integer, Contact> createdContacts = new HashMap<>();
+      contactCsv.forEach((id, name) -> {
+        Contact contact = new Contact();
+        contact.setFirstName(name);
+        contact.setId(id);
+        contact.setCustomer(someCustomer); // customer is mandatory
+        DB.save(contact);
+        createdContacts.put(id, contact);
+      });
+
+
+      contactGroupCsv.forEach((id, name) -> {
+        ContactGroup group = new ContactGroup();
+        group.setName(name);
+        group.setId(id);
+        DB.save(group);
+      });
+
+
+      mappingCsv.forEach((contactId, groupId) -> {
+        Contact contact = createdContacts.get(contactId);
+        contact.setGroup(DB.reference(ContactGroup.class, groupId));
+        DB.save(contact);
```

Comment:
The last code part does some post-processing on some of the objects (the objects may not be saved, yet due ebean batching)

Yes, there are several ways, how to solve this issue
- do a `DB.saveAll(createdContacts.value())`
- disable batching
- do a flush
- write a better import-code 😉

I'm unsure, if ebean can do much here. It must know or adjust the save-order, but this may rise up new problems

Reply:
> I'm unsure, if ebean can do much here

I know its an example but to me its pretty clear that import code should look to first materialise and save the [ContactGroup] beans that will be later _referenced_ by other beans [Contact].

Ebean gives the application **_exact control_** over this. With that control, this import can be done as batch inserts of ContactGroup followed by batch inserts of Contact with no actual updates required [optimal handling of this from the database perspective].

You are going to have a hard time convincing me that ebean should change behaviour here.

What we want is for ebean to give developers the ability to control **_exactly_** how this works because at certain levels of scale we absolutely need that control. Trying to get "fancy" here would not end well.



> do a DB.saveAll(createdContacts.value())

So, do that and rely on ebean cascade persisting from Contact to ContactGroup [and then ebean will determine the ordering because we are now using cascade persist and ebean will get it right]. Yes, a good option.


> disable batching

At some scale people regret this. I say invest the time and think about ordering and stay with jdbc batch [and get orders of magnitude performance benefit from that investment]. I'd personally never go without batching here.


> do a flush

If we really must. Doing so, implies that we'll see _extra_ updates being used when if the ordering was correct the app would avoid those extra updates. So generally not optimal.


> write a better import-code

IMO this is actually the best answer. Actually spend some time thinking about the ordering and design import processing with that in mind, looking to use JDBC batch and avoiding extra updates when possible.